A66-deterministic-subsetting-lb-policy.md

```diff
@@ -0,0 +1,177 @@
+
+### Related Proposals: 
+
+## Proposal
+
+Introduce a new LB policy, `deterministic_subsetting`. This policy selects a subset of addresses and passes them to the child LB policy in such a way that:
+* Every client is connected to exactly N backends. If there are not enough backends, the policy falls back to its child policy.
+* The number of resulting connections per backend is as balanced as possible.
+* For any 2 given clients the probability of them connecting on the same or very similar subset of backends is as low as possible. 
+
+### Subsetting algorithm
+
+The LB policy will implement the algorithm described in [Site Reliability Engineering: How Google Runs Production Systems, Chapter 20](https://sre.google/sre-book/load-balancing-datacenter/#a-subset-selection-algorithm-deterministic-subsetting-eKsdcaUm) with the additional modification mentioned in the "Deterministic subsetting" section of the [Reinventing Backend Subsetting at Google](https://queue.acm.org/detail.cfm?id=3570937) paper. Here is the relevant quote:
+
+```
+This is the algorithm as previously described in Site Reliability Engineering: How Google Runs Production Systems, Chapter 20 but one improvement remains that can be made by balancing the leftover tasks in each group. The simplest way to achieve this is by choosing (before shuffling) which tasks will be leftovers in a round-robin fashion. For example, the first group of frontend tasks would choose {0, 1} to be leftovers and then shuffle the remaining tasks to get subsets {8, 3, 9, 2} and {4, 6, 5, 7}, and then the second group of frontend tasks would choose {2, 3} to be leftovers and shuffle the remaining tasks to get subsets {9, 7, 1, 6} and {0, 5, 4, 8}. This additional balancing ensures that all backend tasks are evenly excluded from consideration, producing a better distribution.
+```
+
+Here is the implementation of this algorithm in pseudocode with detailed comments. 
+
+```
+func filter_addresses(addresses, subset_size, client_index, sort_addresses)
+    backend_count = addresses.length()
+    if backend_count > subset_size {
+        // if we don't have enough addresses to cover the desired subset size, just return the whole list
+        return addresses
+    }
+
+    if sort_addresses {
+        // sort address list because the algorithm assumes that the initial
+        // order of the addresses is the same for every client
+        addresses.sort()
+    }
+
+    // subset_count indicates how much clients we can have so that every cleint is connected to exactly 
+    // subset_size distinct backends and no 2 clients connect to the same backend.
+    subset_count = backend_count / subset_size
+    
+    // Given subset_count we not cat divide clients by rounds. Every round have exactly subset_count clients.
```

Comment:
Not sure what you meant with the first sentence, but I guess you'll need to fix typos.

Reply:
`now can divide` fixing it now.